iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -25,111 +29,230 @@ namespace Flow {
+  if (lhs.getNumOperands() != rhs.getNumOperands()) return false;
+  if (lhs.getNumResults() != rhs.getNumResults()) return false;
+  if (lhs.getNumRegions() != rhs.getNumRegions()) return false;
+  if (lhs.getNumSuccessors() != rhs.getNumSuccessors()) return false;
+
+  // TODO(#3996): symbol mapping; for now allow them to differ unconditionally.
+  if (!compare_ranges(
+          lhs.getAttrs(), rhs.getAttrs(),
+          [&](const NamedAttribute &lhs, const NamedAttribute &rhs) {
+            if (lhs.first == "function_ref" ||
+                lhs.first == SymbolTable::getSymbolAttrName()) {
+              return true;
+            }
+            if (lhs.first != rhs.first) return false;
+            if (lhs.second != rhs.second) return false;
+            return true;
+          })) {
+    return false;
+  }
 
-      // If the operations have regions, recurse into them (depth-first).
-      if (lhsOp.getNumRegions() != rhsOp.getNumRegions()) {
-        return false;
-      }
-      auto lhsRegions = lhsOp.getRegions();
-      auto rhsRegions = rhsOp.getRegions();
-      for (int i = 0; i < lhsRegions.size(); ++i) {
-        if (!areRegionsEquivalent(&lhsRegions[i], &rhsRegions[i])) {
-          return false;
-        }
-      }
+  // If the operation is isolated we don't want to reuse any parent mapping or
+  // pollute it with our mappings.
+  BlockAndValueMapping scopedMapping;
+  BlockAndValueMapping &mapping =
+      lhs.isKnownIsolatedFromAbove() ? scopedMapping : parentMapping;
+
+  // If the op references blocks (such as a branch) then we expect to have them
+  // in the mapping already from the parent region to do the lhs->rhs mapping.
+  for (auto successorPair :
+       llvm::zip(lhs.getSuccessors(), rhs.getSuccessors())) {
+    auto *lhsSuccessor = std::get<0>(successorPair);
+    auto *rhsSuccessor = std::get<1>(successorPair);
+    if (rhsSuccessor != parentMapping.lookup(lhsSuccessor)) return false;
+  }
+
+  // Ensure result types match first and add to the block and value mapping.
+  // For many ops if the result types don't match it's a good (cheap) indicator
+  // that the operands won't match either so this still allows a somewhat-early
+  // exit prior to the full traversal.
+  for (auto resultPair : llvm::zip(lhs.getResults(), rhs.getResults())) {
+    auto &lhsValue = std::get<0>(resultPair);
+    auto &rhsValue = std::get<1>(resultPair);
+    if (lhsValue.getType() != rhsValue.getType()) return false;
+    parentMapping.map(lhsValue, rhsValue);
+  }
+
+  // Check operands using the lhs->rhs mapping; since this op is only consuming
+  // these values they should already be defined in the mapping.
+  for (auto operandPair : llvm::zip(lhs.getOperands(), rhs.getOperands())) {
+    auto &lhsValue = std::get<0>(operandPair);
+    auto &rhsValue = std::get<1>(operandPair);
+    if (lhsValue.getType() != rhsValue.getType()) return false;
+    if (rhsValue != mapping.lookup(lhsValue)) return false;
```

Comment:
This should use parentMapping I think. The operands of an op will be in the parent mapping. (the scoped mapping implied by IsolatedFromAbove goes into effect only inside the child regions)